File: sim/src/distance.py
# distance.py 

import sys
import math

args = sys.argv
args.pop(0) # Ignore self

nbeads = int(args.pop(0))
lx = float(args.pop(0))
ly = float(args.pop(0))
lz = float(args.pop(0))
outfile = args.pop(0)

# compute distance
def dist2(a, b):
    dx = a[0] - b[0]
    dy = a[1] - b[1]
    dz = a[2] - b[2]
    return dx*dx+dy*dy+dz*dz

pos = [[0.0 for j in xrange(3)] for i in xrange(nbeads)]
r2_avg = [0.0 for i in xrange(nbeads)]
#rg2_avg = [0.0 for i in xrange(nbeads)]
r2_count = [0 for i in xrange(nbeads)]

# read x,y,z
for infile in args:
    with open(infile, "r") as f:
        for line in f:
            data = line.strip().split();
            if (len(data) == 9):
                index = int(data[0])
                x = float(data[3])
                y = float(data[4])
                z = float(data[5])
                ix = int(data[6])
                iy = int(data[7])
                iz = int(data[8])
                x += (lx*ix)
                y += (ly*iy)
                z += (lz*iz)
                pos[index-1][0] = x
                pos[index-1][1] = y
                pos[index-1][2] = z

    print "Computing distances for ", infile

    for i in xrange(nbeads):
        cm = [0.0, 0.0, 0.0]
        for j in xrange(i,nbeads):
            sep = abs(j-i)
            r2 = dist2(pos[i],pos[j])
            r2_avg[sep] += r2
            r2_count[sep] += 1

            # Compute cm
"""            cm = [0.0, 0.0, 0.0]
            count = j-i+1
            
            for k in xrange(i,j+1):
                cm[0] += pos[i][0]
                cm[1] += pos[i][1]
                cm[2] += pos[i][2]
            cm[0] /= float(count)
            cm[1] /= float(count)
            cm[2] /= float(count)
            rg2 = 0.0
            for k in xrange(i,j+1):
                rg2 += dist2(pos[k],cm)
            rg2 /= float(count)
            rg2_avg[sep] += rg2
"""                    

# Normalise
for i in xrange(nbeads):
    r2_avg[i] /= float(r2_count[i])
#    rg2_avg[i] /= float(r2_count[i])

with open(outfile, "w") as w:
    for i in xrange(nbeads):
        w.write("{:d} {:.5f}\n".format(i, r2_avg[i]))
#        w.write("{:d} {:.5f} {:.5f}\n".format(i, r2_avg[i], rg2_avg[i]))

File: sim/src/distance.cpp
/* distance.cpp
 * A code to calculate the end-to-end distance using the 
 * final lammps output trajectory file
 */

#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>
#include <memory>
#include <cmath>
#include "Bead.hpp"
#include "Polymer.hpp"
#include "LAMMPS.hpp"

using std::cout;
using std::endl;
using std::string;
using std::ofstream;
using std::shared_ptr;
using std::make_shared;

double dist2(double x, double y, double z);

int main(int argc, char* argv[]) {
  
  if (argc < 6) {
    cout << "Usage: distance [nbeads] [polymerid] " 
	 << "[outfile] [mapfile] [lammpsfile]" << endl;
    return 1;
  }

  int argi {};
  int nbeads {stoi(string(argv[++argi]), nullptr, 10)};
  int polymerid {stoi(string(argv[++argi]), nullptr, 10)};
  string outfile {argv[++argi]};
  string mapfile {argv[++argi]};
  string lammpsfile;
  
  shared_ptr<LAMMPS> lammps = make_shared<LAMMPS>();

  shared_ptr<Bead> bead1, bead2;
  shared_ptr<Polymer> polymer;
  vector<double> r2avg (nbeads, 0.0);
  vector<int> count (nbeads, 0);
  argi++;

  for (; argi < argc; argi++) {
    lammpsfile = string(argv[argi]);
    lammps->importData(lammpsfile, mapfile);
    polymer = lammps->getPolymer(polymerid);

    double lx {lammps->getLx()};
    double ly {lammps->getLy()};
    double lz {lammps->getLz()};

    // Compute distances
    int sep;
    double r2;
    double x, y, z, dx, dy, dz;
    for (int j {}; j < nbeads; j++) {
      bead1 = polymer->getBead(j);
      x = bead1->getPosition(0)+(bead1->getBoundaryCount(0)*lx);
      y = bead1->getPosition(1)+(bead1->getBoundaryCount(1)*ly);
      z = bead1->getPosition(2)+(bead1->getBoundaryCount(2)*lz);
      for (int k {}; k <= j; k++) {
	bead2 = polymer->getBead(k);
	sep = abs(j-k);
	dx = x-(bead2->getPosition(0)+(bead2->getBoundaryCount(0)*lx));
	dy = y-(bead2->getPosition(1)+(bead2->getBoundaryCount(1)*ly));
	dz = z-(bead2->getPosition(2)+(bead2->getBoundaryCount(2)*lz));
	r2 = dist2(dx,dy,dz);
	r2avg[sep] += r2;
	count[sep]++;
      }
    }
  }

  // Normalise results
  for (int i {}; i < nbeads; i++) {
    r2avg[i] /= static_cast<double>(count[i]);
  }
  cout << "Outputting results ..." << endl;
  // Output results
  ofstream writer;
  writer.open(outfile);
  if (!writer) {
    cout << "Problem with opening output file" << endl;
    return 1;
  }
  writer << std::fixed << std::setprecision(5);
  for (int i {}; i < nbeads; i++) {
    writer << i << " " << r2avg[i] << endl;
  }
  writer.close();
}

double dist2(double x, double y, double z) {
  return x*x+y*y+z*z;
}
